File: lib/code_gen/targets/wasm/wasm_translator.rb
# typed: strict
# frozen_string_literal: true

require "sorbet-runtime"
require_relative "../../translator"
require_relative "../../instruction"
require_relative "../../../il"
require_relative "../../../symbol_table"
require_relative "wasm_il_transformer"
require_relative "instructions/instructions"
require_relative "instructions/instruction_set"
require_relative "components"
require_relative "relooper"
require_relative "wasm_block"

module Lilac
  module CodeGen
    module Targets
      module Wasm
        # A WasmTranslator translates Lilac IL into Wasm instructions.
        class WasmTranslator < CodeGen::Translator
          extend T::Sig

          include CodeGen::Targets::Wasm

          sig { params(cfg_program: IL::CFGProgram).void }
          def initialize(cfg_program)
            @symbols = T.let(SymbolTable.new, SymbolTable)
            wasm_transformer = WasmILTransformer.new(@symbols)

            @loop_ct = T.let(0, Integer)
            @block_ct = T.let(0, Integer)

            @start_function = T.let(nil, T.nilable(String))

            super(wasm_transformer, cfg_program)
          end

          sig { params(func_name: String).void }
          # Set the function that will be used as the entry point for the
          # Wasm program. If no start function is set then the Wasm program
          # will have no entry point (no Start component in module).
          def set_start_function(func_name)
            unless @program.get_func(func_name)
              raise "Function '#{func_name}' does not exist in Program."
            end

            @start_function = func_name
          end

          sig { override.returns(Components::Module) }
          def translate
            components = []
            @symbols.push_scope # always have scope for globals and top-level

            # import all extern functions
            @program.each_extern_func do |f|
              components.push(translate_import(f))
            end

            # translate all globals
            @program.each_global do |g|
              # add global to top-level scope
              @symbols.insert(ILSymbol.new(g.id, g.type))
              components.push(translate_global(g))
            end

            # translate all functions
            @program.each_func { |f| components.push(translate_func(f)) }

            # add start component if a start function has been set
            if @start_function
              components.push(Components::Start.new(@start_function))
            end

            Components::Module.new(components)
          end

          private

          sig do
            params(extern_funcdef: IL::ExternFuncDef)
              .returns(Components::Import)
          end
          def translate_import(extern_funcdef)
            # translate param types for function signature
            param_types = []
            extern_funcdef.param_types.each do |t|
              param_types.push(Instructions.to_wasm_type(t))
            end

            results = []
            unless extern_funcdef.ret_type.is_a?(IL::Types::Void)
              results.push(Instructions.to_wasm_type(extern_funcdef.ret_type))
            end

            # construct import object
            Components::Import.new(extern_funcdef.source,
                                   extern_funcdef.name,
                                   param_types,
                                   results)
          end

          sig { params(global_def: IL::GlobalDef).returns(Components::Global) }
          def translate_global(global_def)
            type = Instructions.to_wasm_type(global_def.type)
            name = global_def.id.name
            const = @transformer.transform(global_def.rhs)[0]

            unless const.is_a?(Instructions::ConstInteger) ||
                    const.is_a?(Instructions::ConstFloat)
              raise "RHS of GlobalDef did not translate to a ConstInteger or "\
                    "a ConstFloat"
            end

            # TODO: optimization to make some globals not mutable
            Components::Global.new(type, name, const, mutable: true)
          end

          sig { params(cfg_funcdef: IL::CFGFuncDef).returns(Components::Func) }
          def translate_func(cfg_funcdef)
            params = [] # list of params for function signature

            # create a new scope with func params
            @symbols.push_scope
            cfg_funcdef.params.each do |p|
              @symbols.insert(ILSymbol.new(p.id, p.type))

              # also mark this down for the function signature
              param_type = Instructions.to_wasm_type(p.type)
              param_name = p.id.name
              params.push(Components::Local.new(param_type, param_name))
            end

            # find all locals in the function
            locals = find_locals(cfg_funcdef.cfg)

            # translate instructions and pop the scope we used
            instructions = translate_instructions(cfg_funcdef.cfg)
            @symbols.pop_scope
            instructions.push(Instructions::End.new) # every func ends with End

            # construct func with appropriate params and return type
            # if return type is void then result is simply nil
            results = []
            unless cfg_funcdef.ret_type.is_a?(IL::Types::Void)
              results.push(Instructions.to_wasm_type(cfg_funcdef.ret_type))
            end
            export_name = cfg_funcdef.exported ? cfg_funcdef.name : nil
            Components::Func.new(cfg_funcdef.name,
                                 params,
                                 results,
                                 locals,
                                 instructions,
                                 export: export_name)
          end

          sig do
            params(cfg: Analysis::CFG)
              .returns(T::Hash[Type, T::Array[Components::Local]])
          end
          def find_locals(cfg)
            locals = {}

            # scan forwards to build a symbol table of all locals and their
            #   types.
            # NOTE: assume that our caller has kindly pushed a new scope
            cfg.each_node do |b|
              # scan each block for definitions and log them
              # a previous validation should have already ensured that there are
              # no inconsistent types, so we don't need any checks here.
              b.stmt_list.each do |s|
                unless s.is_a?(IL::Definition)
                  next
                end

                # avoid redefining globals, params, or ids used more than once
                if @symbols.lookup(s.id)
                  next
                end

                # log in symbol table
                id_symbol = ILSymbol.new(s.id, s.type)
                @symbols.insert(id_symbol)

                # push declaration
                type = Instructions.to_wasm_type(s.type)
                new_local = Components::Local.new(type, s.id.name)
                if locals[type]
                  locals[type].push(new_local)
                else
                  locals[type] = [new_local]
                end
              end
            end

            locals
          end

          sig do
            params(cfg: Analysis::CFG)
              .returns(T::Array[Instructions::WasmInstruction])
          end
          def translate_instructions(cfg)
            # create a relooper instance for the cfg
            relooper = Relooper.new(cfg)

            instructions = []

            # run relooper, translate the WasmBlocks, and then return that
            root = relooper.translate
            instructions.concat(translate_wasm_block(root))

            instructions
          end

          sig do
            params(block: WasmBlock)
              .returns(T::Array[Instructions::WasmInstruction])
          end
          def translate_wasm_block(block)
            case block
            when WasmIfBlock
              instructions = []

              # translate all the conditional stuff
              block.bb.stmt_list.each do |s|
                instructions.concat(@transformer.transform(s))
              end

              # push the conditional and create the if
              bb_exit = T.unsafe(block.bb.exit)
              unless bb_exit
                raise "Basic block did not have a exit"
              end

              instructions.push(push_value(bb_exit.cond))
              # TODO: handle jump_zero with non-int conditional value
              if bb_exit.is_a?(IL::JumpZero)
                cond_il_type = T.cast(@transformer, WasmILTransformer)
                                .get_il_type(bb_exit.cond)

                # handle integer or non-integer IL type
                if cond_il_type.is_a?(IL::Types::IntegerType)
                  cond_type = Instructions.to_integer_type(cond_il_type)
                  instructions.push(Instructions::EqualZero.new(cond_type))
                elsif cond_il_type.is_a?(IL::Types::FloatType)
                  cond_type = Instructions.to_float_type(cond_il_type)
                  instructions.push(
                    Instructions::ConstFloat.new(cond_type, "0.0")
                  )
                  instructions.push(Instructions::Equal.new(cond_type))
                else # NOTE: I think this never happens
                  raise "Unexpected conditional IL type"
                end
              end
              instructions.push(Instructions::If.new)

              # no true branch = invalid WasmIfBlock
              unless block.true_branch
                raise "WasmIfBlock has no true branch"
              end

              # translate true branch
              if_true_branch = translate_wasm_block(T.unsafe(block.true_branch))
              instructions.concat(if_true_branch)

              # fill in EmptyType if true branch is empty
              if if_true_branch.empty?
                instructions.push(Instructions::EmptyType.new)
              end

              # false branch is optional
              if block.false_branch
                # create the else
                instructions.push(Instructions::Else.new)

                # translate false branch
                block_false_branch = T.unsafe(block.false_branch)
                if_false_branch = translate_wasm_block(block_false_branch)
                instructions.concat(if_false_branch)

                # fill in EmptyType if false branch is empty
                if if_false_branch.empty?
                  instructions.push(Instructions::EmptyType.new)
                end
              end

              # end the if else statement
              instructions.push(Instructions::End.new)

              # translate the following blocks
              if block.next_block
                block_next_block = T.unsafe(block.next_block)
                instructions.concat(translate_wasm_block(block_next_block))
              end

              instructions
            when WasmLoopBlock
              instructions = []
              block_label = alloc_block_label
              loop_label = alloc_loop_label

              # translate all the conditional stuff
              # this will be placed at the beginning of the _block_ and
              # the end of the _loop_
              cond_insts = []
              block.bb.stmt_list.each do |s|
                cond_insts.concat(@transformer.transform(s))
              end
              bb_exit = T.unsafe(block.bb.exit)
              unless bb_exit
                raise "Basic block did not have a exit"
              end

              cond_insts.push(push_value(bb_exit.cond))
              if bb_exit.is_a?(IL::JumpZero) # TODO: handle JumpNotZero?
                cond_il_type = T.cast(@transformer, WasmILTransformer)
                                .get_il_type(bb_exit.cond)

                # handle integer or non-integer IL type
                if cond_il_type.is_a?(IL::Types::IntegerType)
                  cond_type = Instructions.to_integer_type(cond_il_type)
                  cond_insts.push(Instructions::EqualZero.new(cond_type))
                elsif cond_il_type.is_a?(IL::Types::FloatType)
                  cond_type = Instructions.to_float_type(cond_il_type)
                  cond_insts.push(
                    Instructions::ConstFloat.new(cond_type, "0.0")
                  )
                  cond_insts.push(Instructions::Equal.new(cond_type))
                else # NOTE: I think this never happens
                  raise "Unexpected conditional IL type"
                end
              end

              # create the outer block
              instructions.push(Instructions::Block.new(block_label))

              # conditional check before entering the loop
              # (since this is emulating a while, not a do-while)
              instructions.concat(cond_insts)
              # push conditional branch
              instructions.push(Instructions::BranchIf.new(block_label))

              # create the loop
              instructions.push(Instructions::Loop.new(loop_label))

              # translate the inner of the loop (not optional)
              unless block.inner
                raise "WasmLoopBlock has no inner"
              end

              inner = translate_wasm_block(T.unsafe(block.inner))
              instructions.concat(inner)

              # conditional check at end of the loop to see if we should exit
              instructions.concat(cond_insts)
              instructions.push(Instructions::BranchIf.new(block_label))

              # unconditional jump back after the check
              instructions.push(Instructions::Branch.new(loop_label))

              # end the loop and the block
              instructions.push(Instructions::End.new)
              instructions.push(Instructions::End.new)

              # translate the following blocks
              if block.next_block
                block_next_block = T.unsafe(block.next_block)
                instructions.concat(translate_wasm_block(block_next_block))
              end

              instructions
            when WasmBlock
              instructions = []

              # translate block
              block.bb.stmt_list.each do |s|
                instructions.concat(@transformer.transform(s))
              end

              # translate next block
              if block.next_block
                block_next_block = T.unsafe(block.next_block)
                instructions.concat(translate_wasm_block(block_next_block))
              end

              instructions
            end
          end

          sig do
            params(value: IL::Value).returns(Instructions::WasmInstruction)
          end
          def push_value(value)
            case value
            when IL::Constant
              if value.type.is_a?(IL::Types::IntegerType)
                Instructions::ConstInteger.new(
                  Instructions.to_integer_type(value.type),
                  value.value
                )
              elsif value.type.is_a?(IL::Types::FloatType)
                Instructions::ConstFloat.new(
                  Instructions.to_float_type(value.type),
                  value.value
                )
              end

              raise "IL::Constant is neither integer nor float"
            when IL::ID
              Instructions::LocalGet.new(value.name)
            when IL::Value # cannot happen, IL::Value is abstract
              raise "Attempted to push value of stub IL::Value"
            else
              T.absurd(value)
            end
          end

          sig { returns(String) }
          def alloc_block_label
            label = "__lilac_block#{@block_ct}"
            @block_ct += 1
            label
          end

          sig { returns(String) }
          def alloc_loop_label
            label = "__lilac_loop_#{@loop_ct}"
            @loop_ct += 1
            label
          end
        end
      end
    end
  end
end

File: Lilac/src/CodeGen/Targets/Wasm/WasmTranslator.cs
using Lilac.Analysis;
using Lilac.CodeGen.Targets.Wasm.Instructions;
using Lilac.IL;
using Type = Lilac.CodeGen.Targets.Wasm.Instructions.Type;

namespace Lilac.CodeGen.Targets.Wasm;

public class WasmTranslator : Translator<WasmInstruction> {
  private SymbolTable symbols;

  private int loopCt;
  private int blockCt;

  private string? _startFunction;
  public string? StartFunction {
    get => _startFunction;
    set {
      _startFunction = value;
      if (CFGProgram.GetFunc(_startFunction) is null) {
        throw new KeyNotFoundException();
      }
    }
  }

  public WasmTranslator(CFGProgram cfgProgram) {
    symbols = new();

    loopCt = 0;
    blockCt = 0;

    StartFunction = null;

    Transformer = new WasmILTransformer(symbols);
    CFGProgram = cfgProgram;
  }

  public override Module Translate() {
    List<WasmComponent> components = [];
    symbols.PushScope(); // always have scope for globals and top-level

    // import all extern functions
    foreach (ExternFuncDef f in CFGProgram.GetExternFuncs()) {
      components.Add(TranslateImport(f));
    }

    // translate all globals
    foreach (GlobalDef g in CFGProgram.GetGlobals()) {
      // add global to top-level scope
      symbols.Add(new(g.Id, g.Type));
      components.Add(TranslateGlobal(g));
    }

    // translate all functions
    foreach (CFGFuncDef f in CFGProgram.GetFuncs()) {
      components.Add(TranslateFunc(f));
    }

    // add start component if a start function has been set
    if (StartFunction is not null) {
      components.Add(new Start(StartFunction));
    }

    return new Module(components);
  }

  private Import TranslateImport(ExternFuncDef externFuncDef) {
    // translate param types for function signature
    List<Type> paramTypes = [];
    foreach (IL.Type t in externFuncDef.ParamTypes) {
      paramTypes.Add(t.ToWasmType());
    }

    List<Type> results = [];
    if (!externFuncDef.RetType.IsVoid()) {
      results.Add(externFuncDef.RetType.ToWasmType());
    }

    return new Import(externFuncDef.Source,
                      externFuncDef.Name,
                      paramTypes,
                      results);
  }

  private Global TranslateGlobal(GlobalDef globalDef) {
    Type type = globalDef.Type.ToWasmType();
    string name = globalDef.Id.Name;
    List<WasmInstruction> rhs = Transformer.Transform(globalDef.Rhs);

    if (rhs.Count == 0 || rhs[0] is not Const) {
      throw new Exception("RHS of GlobalDef did not translate to a Wasm Const");
    }

    // TODO: optimization to make some globals not mutable
    return new Global(type, name, (Const)rhs[0], true);
  }

  private Func TranslateFunc(CFGFuncDef cfgFuncDef) {
    List<Local> @params = []; // list of params for functions signature

    // create a new scope with func params
    symbols.PushScope();
    foreach (FuncParam p in cfgFuncDef.Params) {
      symbols.Add(new(p.Id, p.Type));

      // also mark this down for the function signature
      Type paramType = p.Type.ToWasmType();
      string paramName = p.Id.Name;
      @params.Add(new Local(paramType, paramName));
    }

    // find all locals in the function
    Dictionary<Type, List<Local>> locals = FindLocals(cfgFuncDef.CFG);

    // translate instructions and pop the scope we used
    List<WasmInstruction> instructions = TranslateInstructions(cfgFuncDef.CFG);
    symbols.PopScope();
    instructions.Add(new End()); // every func ends with End

    // construct func with appropriate params and return type
    // if return type is void then result is simply null
    List<Type> results = [];
    if (!cfgFuncDef.RetType.IsVoid()) {
      results.Add(cfgFuncDef.RetType.ToWasmType());
    }

    string? exportName = cfgFuncDef.Exported ? cfgFuncDef.Name : null;
    return new Func(cfgFuncDef.Name,
                    @params,
                    results,
                    locals,
                    instructions,
                    exportName);
  }

  private Dictionary<Type, List<Local>> FindLocals(CFG cfg) {
    Dictionary<Type, List<Local>> locals = new();

    // scan forwards to build a symbol table of all locals and their types
    // NOTE: assume that our caller has pushed a new scope for us
    foreach (BB n in cfg.GetNodes()) {
      // scan each block for definitions and log them
      // assume that there are no inconsistent types
      foreach (Statement s in n.StmtList) {
        if (s is not Definition d) {
          continue;
        }

        // avoid redefining globals, params, or ids used more than once
        if (symbols[d.Id] is not null) {
          continue;
        }

        // log in symbol table
        symbols.Add(new(d.Id, d.Type));

        // push declaration
        Type type = d.Type.ToWasmType();
        Local newLocal = new(type, d.Id.Name);

        if (!locals.TryGetValue(type, out List<Local>? typeList)) {
          locals.Add(type, [newLocal]);
        }
        else {
          typeList.Add(newLocal);
        }
      }
    }

    return locals;
  }

  private List<WasmInstruction> TranslateInstructions(CFG cfg) {
    // create a relooper instance for the cfg
    Relooper relooper = new(cfg);

    List<WasmInstruction> instructions = [];

    // run relooper, translate the WasmBlocks, and then return that
    WasmBlock root = relooper.Translate();
    instructions.AddRange(TranslateWasmBlock(root));

    return instructions;
  }

  private List<WasmInstruction> TranslateWasmBlock(WasmBlock block) {
    switch (block) {
      case WasmIfBlock ifBlock: {
        List<WasmInstruction> instructions = [];

        // translate all the conditional stuff
        foreach (Statement s in ifBlock.BB.StmtList) {
          instructions.AddRange(Transformer.Transform(s));
        }

        // push the conditional and create the if
        CondJump? bbExit = ifBlock.BB.Exit as CondJump;
        if (bbExit is null) {
          throw new Exception("Basic block did not have an exit");
        }

        instructions.Add(PushValue(bbExit.Cond));
        if (bbExit is JumpZero) {
          IL.Type condIlType =
            ((WasmILTransformer)Transformer).GetILType(bbExit.Cond);

          // handle integer or non-integer IL type
          if (condIlType.IsInteger()) {
            instructions.Add(new EqualZero(condIlType.ToWasmType()));
          }
          else if (condIlType.IsFloat()) {
            Type type = condIlType.ToWasmType();
            instructions.Add(new Const(type, "0.0"));
            instructions.Add(new Equal(type));
          }
          else {
            throw new ArgumentOutOfRangeException();
          }
        }

        instructions.Add(new If());

        // no true branch = invalid WasmIfBlock
        if (ifBlock.TrueBranch is null) {
          throw new NullReferenceException();
        }

        // translate true branch
        List<WasmInstruction> ifTrueBranch =
          TranslateWasmBlock(ifBlock.TrueBranch);
        instructions.AddRange(ifTrueBranch);

        // fill in empty type if true branch is empty
        if (ifTrueBranch.Count == 0) {
          instructions.Add(new EmptyType());
        }

        // false branch if optional
        if (ifBlock.FalseBranch is not null) {
          // create the else
          instructions.Add(new Else());

          // translate false branch
          List<WasmInstruction> ifFalseBranch =
            TranslateWasmBlock(ifBlock.FalseBranch);
          instructions.AddRange(ifFalseBranch);

          // fill in empty type if false branch is empty
          instructions.Add(new EmptyType());
        }

        // end the if statement
        instructions.Add(new End());

        // translate the next blocks
        if (block.NextBlock is not null) {
          instructions.AddRange(TranslateWasmBlock(block.NextBlock));
        }

        return instructions;
      }
      case WasmLoopBlock loopBlock: {
        List<WasmInstruction> instructions = [];
        string blockLabel = AllocBlockLabel();
        string loopLabel = AllocLoopLabel();

        // translate all the conditional stuff
        // this will be placed at the beginning of the _block_ and the end of
        // the _loop_
        List<WasmInstruction> condInsts = [];
        foreach (Statement s in loopBlock.BB.StmtList) {
          condInsts.AddRange(Transformer.Transform(s));
        }

        CondJump? bbExit = loopBlock.BB.Exit as CondJump;
        if (bbExit is null) {
          throw new NullReferenceException();
        }

        condInsts.Add(PushValue(bbExit.Cond));
        if (bbExit is JumpZero) { // TODO: handle jnz?
          IL.Type condIlType =
            ((WasmILTransformer)Transformer).GetILType(bbExit.Cond);

          // handle integer or non-integer IL type
          if (condIlType.IsInteger()) {
            instructions.Add(new EqualZero(condIlType.ToWasmType()));
          }
          else if (condIlType.IsFloat()) {
            Type type = condIlType.ToWasmType();
            instructions.Add(new Const(type, "0.0"));
            instructions.Add(new Equal(type));
          }
          else {
            throw new ArgumentOutOfRangeException();
          }
        }

        // create the outer block
        instructions.Add(new Block(blockLabel));

        // conditional check before entering the loop
        // (since this is emulating a while, not a do-while)
        instructions.AddRange(condInsts);

        // push conditional branch
        instructions.Add(new BranchIf(blockLabel));

        // create the loop
        instructions.Add(new Loop(loopLabel));

        // translate the inner of the loop (not optional)
        if (loopBlock.Inner is null) {
          throw new NullReferenceException();
        }

        instructions.AddRange(TranslateWasmBlock(loopBlock.Inner));

        // conditional check at end of the loop to see if we should exit
        instructions.AddRange(condInsts);
        instructions.Add(new BranchIf(blockLabel));

        // unconditional jump back after the check
        instructions.Add(new Branch(loopLabel));

        // end the loop and the block
        instructions.Add(new End());
        instructions.Add(new End());

        // translate the next blocks
        if (block.NextBlock is not null) {
          instructions.AddRange(TranslateWasmBlock(block.NextBlock));
        }

        return instructions;
      }
      case WasmBlock: {
        List<WasmInstruction> instructions = [];

        // translate block
        foreach (Statement s in block.BB.StmtList) {
          instructions.AddRange(Transformer.Transform(s));
        }

        // translate the next blocks
        if (block.NextBlock is not null) {
          instructions.AddRange(TranslateWasmBlock(block.NextBlock));
        }

        return instructions;
      }
      default:
        throw new ArgumentOutOfRangeException();
    }
  }

  private WasmInstruction PushValue(Value value) {
    return value switch {
      Constant constant when !constant.Type.IsVoid() =>
        new Const(constant.Type.ToWasmType(),
                  constant.Value.ToString()?? string.Empty),
      Constant => throw new Exception("IL Constant is Void"),
      GlobalID id => new GlobalGet(id.Name),
      ID id => new LocalGet(id.Name),
      _ => throw new ArgumentOutOfRangeException(),
    };
  }

  private string AllocBlockLabel() {
    string label = $"__block_{blockCt}";
    blockCt += 1;
    return label;
  }

  private string AllocLoopLabel() {
    string label = $"__loop_{loopCt}";
    loopCt += 1;
    return label;
  }
}